server/client.go

```diff
@@ -2132,7 +2132,7 @@ func (c *client) processPong() {
 
 // Will return the parts from the raw wire msg.
 func (c *client) msgParts(data []byte) (hdr []byte, msg []byte) {
-	if c != nil && c.pa.hdr > 0 {
+	if c != nil && c.pa.hdr > 0 && len(data) > 0 {
```

Comment:
@derekcollison Well, my understanding is that this is worrisome because if `c.pa.hdr > 0` then data should not be empty because data should contain both header and payload.
I think the issue is (o *consumer) processAck() that calls processNextMsgReq() with msg[len(AckNext):]. That is wrong, we need to differentiate processNextMsgReq() when coming from processAck and from subscribeInternal().

Reply:
Correct something is off there, this is more of a bandaid.

Reply:
specifically, this is happening when sending AckNext on a message that was pulled from another account as in this test: https://github.com/nats-io/nats.go/blob/master/test/js_test.go#L1729-L1730

Reply:
I would propose doing that instead:
```
diff --git a/server/client.go b/server/client.go
index 0055f170..bb563c46 100644
--- a/server/client.go
+++ b/server/client.go
@@ -2132,7 +2132,7 @@ func (c *client) processPong() {
 
 // Will return the parts from the raw wire msg.
 func (c *client) msgParts(data []byte) (hdr []byte, msg []byte) {
-       if c != nil && c.pa.hdr > 0 && len(data) > 0 {
+       if c != nil && c.pa.hdr > 0 {
                return data[:c.pa.hdr], data[c.pa.hdr:]
        }
        return nil, data
diff --git a/server/consumer.go b/server/consumer.go
index 33ea9be3..18507c66 100644
--- a/server/consumer.go
+++ b/server/consumer.go
@@ -1036,7 +1036,10 @@ func (o *consumer) processAck(_ *subscription, c *client, subject, reply string,
                o.ackMsg(sseq, dseq, dc)
        case bytes.HasPrefix(msg, AckNext):
                o.ackMsg(sseq, dseq, dc)
+               phdr := c.pa.hdr
+               c.pa.hdr = -1
                o.processNextMsgReq(nil, c, subject, reply, msg[len(AckNext):])
+               c.pa.hdr = phdr
                skipAckReply = true
        case bytes.Equal(msg, AckNak):
                o.processNak(sseq, dseq)
```
I believe that processNextMsgReq() ignores the header after calling msgParts(), so just let trick it into thinking there is no header when called from processAck?